### PR TITLE
Phase-hgx323Again Further correction for ring number assignment of scintillator hits in HGCal geometry for versions beyond V13

### PR DESCRIPTION
**Pull request by bsunanda**
#### PR description:

Further correction for ring number assignment of scintillator hits in HGCal geometry versions V14, V15, V16 and V17. This new PR was made because of merging #39164 and then reverting #39198. I wonder why?

#### PR validation:

Use the runTheMatrix test workflows

#### If this PR is a backport please specify the original PR and why you need to backport that PR. If this PR will be backported please specify to which release cycle the backport is meant for:

This is a bug fix for all versions of HGCal geometry

### CODE REVIEW

**Comment by bsunanda** [comment]
@cmsbuild Please test

File: Geometry/HGCalCommonData/src/HGCalDDDConstants.cc [hunk 1]
```diff
@@ -590,13 +590,14 @@ bool HGCalDDDConstants::isValidHex8(int layer, int modU, int modV, int cellU, in
...
+bool HGCalDDDConstants::isValidTrap(int zside, int layer, int irad, int iphi) const {
   // Check validity for a layer|eta|phi of scintillator
   const auto& indx = getIndex(layer, true);
   if (indx.first < 0)
     return false;
-  return ((irad >= hgpar_->iradMinBH_[indx.first]) && (irad <= (hgpar_->iradMaxBH_[indx.first] + 1)) && (iphi > 0) &&
-          (iphi <= hgpar_->scintCells(layer)));
+  bool ok = ((irad >= hgpar_->iradMinBH_[indx.first]) && (irad <= (hgpar_->iradMaxBH_[indx.first] + 1)) && (iphi > 0) &&
+             (iphi <= hgpar_->scintCells(layer)));
+  return ((ok && trapezoidFile()) ? tileExist(zside, layer, irad, iphi) : ok);
```
Comment (perrotta) [inline Geometry/HGCalCommonData/src/HGCalDDDConstants.cc:600]:
```suggestion
  return ((ok && trapezoidFile()) ? tileExist(zside, layer, irad, iphi) : true);
```
I understand that in practice `ok && ok` is identical to `ok && true`. But translating in C++ the sentence "irrespective of trapezoidFile() the earlier test is good for all types of scintillators" is exactly `ok && true`

Comment (bsunanda) [reply Geometry/HGCalCommonData/src/HGCalDDDConstants.cc:600]: It should return the results ok if it is false and if ok is true, then one should see if TrapezoidFile() is the case then should return a test on tileExist. That is the correct logic

**Comment by bsunanda** [comment]
@cmsbuild type bug-fix

**Comment by bsunanda** [comment]
type bug-fix

**Comment by srimanob** [comment]
+Upgrade

Change in D77 is expected (see comment https://github.com/cms-sw/cmssw/pull/39164#issuecomment-1227920299). Change in D88 is unclear if it will be visible, this should be followed up with HGCAL. Propose to merge this PR to `CMSSW_12_5_0_pre5`, then follow up on HGCAL validation. FYI @cms-sw/hgcal-dpg-l2 

Copy the comment from original PR #39164
This bug was there all the time. When the position at the border of 2 rings, it is usually referred to as the lower ring number. This is fine as long as there is no tile in the lower ring. We did not have a good valid test of a detid earlier. We have to introduce one now and that revealed the discrepancy. We do have detailed testing code in the repository and we do conduct them to look into details. The standard validation code can look into broad discrepancies so far and the detailed tests are better done by the experts of the geometry code.

**Comment by perrotta** [comment]
+1
- Let have this merged for next pre5 (@smuzaffar kindly agreed to postpone the 1100 IB for it)
- Since @cms-sw/geometry-l2 and @cms-sw/simulation-l2 already signed the identical https://github.com/cms-sw/cmssw/pull/39164, I assume I can take their signature for granted also here
- Differences in D77 outpus are expected, see also https://github.com/cms-sw/cmssw/pull/39206#issuecomment-1228268005: a more detailed comparison will be possible with the 12_5_0_pre5 phase2 relval campaigns
- Issues reported by the SA are independent from this PR, and they are anyhow dealt woth separately in #39208 and #39209

**Comment by perrotta** [comment]
merge

**Comment by bsunanda** [comment]
+geometry

**Pull request merged**